add fish completions support

isaec opened the pull request:
closes https://projects.torsion.org/borgmatic-collective/borgmatic/issues/686

I've tested this on my system and it seems to work well. There is room to improve on the completions, but parsing the args is a little tedious, and this works well enough for my use.

I don't write python, so I'm sorry if this isn't pythonic.

witten reviewed (changes requested):
This looks great! Thank you so much for taking this on and updating the docs, too.

The one thing I see as missing is automated testing. Fortunately, the level of testing done for the Bash completions and needed (IMO) for Fish completions is super minimal and high-level. For instance, see `tests/integration/commands/test_completion.py` and `tests/end-to-end/test_completion.py`. They test, respectively, that the completions: 1. Don't raise (integration test), and 2. Actually get interpreted as valid Bash (end-to-end) test.

For adding the latter for Fish, you'll need to edit line 20-21 of `scripts/run-full-tests` and add `fish` to the `apk add` command so that it actually gets installed for use in end-to-end tests.

isaec commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +58,55 @@ def bash_completion():
+            for actionStr, subparser in subparsers.choices.items()
+            for action in subparser._actions
+        )
+    )
```
yep! its not quite as good as I would like it to be, but I wasn't able to combine it to group equivalent arguments. Fish doesn't seem to really appreciate the "everything is actually an argument" approach to argparsing, it really wants them to be more like flags associated with args.

However, it has the correct arguments with descriptions, and only shows the contextual args contextually, so that's pretty decent.
![scr_13:11:10](https://user-images.githubusercontent.com/72410860/236318599-09844544-421f-4639-b075-8de2c526ee3f.png)
![image](https://user-images.githubusercontent.com/72410860/236318548-10054fbe-31fd-41ed-9788-063feca4360d.png)

In particular, knowing when file completions are correct seems to be a bit busted...
![image](https://user-images.githubusercontent.com/72410860/236318831-bbfb1eb5-d462-4dd0-a6f5-f2c76774850b.png)
I can try and fix that, it seems like the conditional arg gets associated in some way.

witten replied:
Pretty cool! Much nicer than the Bash completions, at least the way it's presented to the user. It's up to you whether it's worth sinking more time into the file completions or not.

isaec commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +58,55 @@ def bash_completion():
+            '\n# subparser flags',
+        ) + tuple(
+            '''complete -c borgmatic -a '%s' -d %s -n "__fish_seen_subcommand_from %s" %s'''
+            % (' '.join(action.option_strings), shlex.quote(action.help), actionStr, build_fish_flags(action))
```
yep, that would work, I think I got frustrated with trying to put quotes inside fstrings or escape characters and gave up, but the final iteration should be fine with fstrings. I'll make these changes.

isaec commented:
I went a little overboard, and fixed the file completions as well as adding support for the choice style completions, and requiring a value when nargs dictates it. I think I resolved all the feedback on the first draft (fish v Fish, unit and integration tests, fstrings) although I doubtless made some new messes.
![image](https://user-images.githubusercontent.com/72410860/236391292-052f7982-2cbb-4c3c-9a88-951948c466e7.png)
![image](https://user-images.githubusercontent.com/72410860/236391351-34fb3ec6-a655-4e96-a618-45b6d3ee98b5.png)
![image](https://user-images.githubusercontent.com/72410860/236391401-3c27b830-93d8-418d-b808-54b429076a4e.png)

The unit tests pass on my system and I can see the new fish integration and unit test in the output, so I figure it's working. It might be interesting to call ``complete -C`` in the integration test to see if the completions are actually being added, but this is likely not a useful test.

Getting the more powerful contextual completions right in a few cases required sorta targeted (and thus brittle) checks, although I think they would degrade gracefully.

I considered archive selection having the possible archives, but I think this is a can of worms on systems like mine where borgmatic can only be called as sudo and completions don't execute as sudo.

I also made the file match check async, which reduces the delay in initial completion display by ~100ms on my system.

Thank you for your wonderful project and your time reviewing my PR!

witten commented:
Wow, super cool! This now goes way further than the existing Bash completion code, which does file completions but that's pretty much it when it comes to flag values.

> I considered archive selection having the possible archives, but I think this is a can of worms on systems like mine where borgmatic can only be called as sudo and completions don't execute as sudo.

Yeah, that seems potentially thorny to get right.

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+    return action.metavar in (
+        'FILENAME',
+        'PATH',
+    ) or action.dest in ('config_paths',)
```
Nice! I agree that this is probably fine, especially given that nothing blows up if there's a false negative.

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+            '+',
+            '*',
+        )
+        or '--archive' in action.option_strings
```
I noticed that above you do `action.dest in ('config_paths',)`, but here you check the `option_strings`. However it might be nice to pick one approach and do it everywhere you're trying to identify a particular command-line flag.

Do not feel strongly.

witten replied:
Also, what's the purpose of this particular check? Just to match the couple of actions which have an optional rather than required `--archive` flag? But then isn't that kind of violating the name of the function.. `has_required_param_options()`? (This is definitely a case where a docstring for the function would help me out!)

isaec replied:
The purpose is to prevent the normal completions from showing up when you hit tab. Given the default is already set for archive, its sorta weird to use it without specifying an archive... but I removed it. Much more helpful if the completions could know about archives. Could borgmatic perhaps have a method that returns a cached list of archives from the last time it interacted with the server, that doesn't require auth?

witten replied:
Thanks for the explanation! A cached list of archives would be possible—but probably not as part of this initial ticket. Feel free to file a new one for that feature though. It might be a little tricky because borgmatic doesn't (currently) list archives upon create. Although I guess it could implicitly append any newly created archive names to the cache without having to run `list`...

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+            '*',
+        )
+        or '--archive' in action.option_strings
+        or action.metavar in ('PATTERN', 'KEYS', 'N')
```
Similar here. This looks like it's matching optional flags, which I'm assuming _is_ the intent because you went to the trouble of picking them out here.. I just don't know why.

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+    '''
+    Given an argparse.Action instance, return a completion invocation
+    that forces file completion or options completion, if the action
+    takes such an argument and was the last action on the command line.
```
FYI you can wrap at 100 characters in this codebase.

isaec replied:
Is there a way to automate this here? The formatters don't seem to do it and I don't have a good eye for character count.

witten replied:
Depends on your IDE.

 * VSCode: https://stackoverflow.com/questions/43122175/automatically-hard-wrap-lines-at-column-in-vscode
 * vim: https://vi.stackexchange.com/questions/18798/automatic-wraping-to-new-line

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+    if has_required_param_options(action):
+        return f'''\ncomplete -c borgmatic -x -n "__borgmatic_last_arg {args}"'''
+
+    raise RuntimeError(
```
`ValueError` might be a little more specific (and still appropriate) here.

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+        return f'''\ncomplete -c borgmatic -x -n "__borgmatic_last_arg {args}"'''
+
+    raise RuntimeError(
+        f'Unexpected action: {action} passes has_exact_options but has no choices produced'
```
Given that this error message could make its way to a user, it might be a good idea to make it somewhat more user-centric (at the cost of specificity). E.g.: `Could not produce a tab completion for action {action}` or whatever. You can always include more information for the hapless debugging dev with a comment.

isaec replied:
can this error make it to a user? The unit test should fail and as far as I am aware actions are not created on the fly, I think this serves more as an impossible case where you forget to handle one of the cases you filtered for.

witten replied:
Okay, got it.

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+
+
+def dedent_strip_as_tuple(string: str):
+    return (dedent(string).strip('\n'),)
```
So here's another thing that might seem like overkill: Even though all of these nicely separated utility functions are covered by the big ol' integration / end-to-end tests you added, IMO they should also have unit test coverage. Part of that is because they _are_ so nicely separated and atomic—you can pretty easily pass in a single value (or a mock) and test that each function returns what it's supposed to.

In cases where a function really has no logic to speak of, a simple "does not raise" unit test would be sufficient IMO.

And I still believe though that you don't need any unit testing on `fish_completion()` itself. The cost/benefit there just wouldn't be worth it.

isaec replied:
I think I did this? Not sure if the tests I wrote were quite what you were looking for.

witten commented on borgmatic/commands/completion.py:
```diff
@@ -55,3 +65,153 @@ def bash_completion():
+            end
+            __borgmatic_check_version
+
+            function __borgmatic_last_arg --description 'Check if any of the given arguments are the last on the command line'
```
What if someone is trying to do a completion in the middle of a line?

isaec replied:
![image](https://user-images.githubusercontent.com/72410860/236618452-d3389b38-1ec8-497a-8d29-728b9c3ba224.png)
Great question, works fine, but that means the comment is wrong--should say prior to cursor or something... the specific flags passed to commandline ignore anything after the cursor.

witten reviewed (changes requested):
The overall approach looks good to me. I'm envious about these choice completions and may have to steal that for Bash! I think my main feedback is that I'd love to see unit tests and docstrings for the helpful utility functions you added. I'd be happy to help answer any questions about testing those if you're up to doing it.

isaec commented:
Renamed the PR because these aren't basic anymore :laughing: 
I added unit tests, although they don't look quite like the style you have elsewhere.
I added some doccomments although they might not be clear enough...

witten commented on tests/unit/commands/test_completions.py:
```diff
@@ -0,0 +1,143 @@
+
+import pytest
+
+from borgmatic.commands.completion import (
```
So the convention in other tests in borgmatic would be to do this instead:

```python
from borgmatic.commands import completion as module
```

And then wherever you need to refer to a unit under test, it would look like `module.dedent_strip_as_tuple` or `module.exact_options_completion`.

The intent of this convention, BTW, is that it makes it real easy to identify the unit under test when reading a test function.

witten commented on tests/unit/commands/test_completions.py:
```diff
@@ -0,0 +1,143 @@
+            OptionType(file=True, choice=False, unknown_required=False),
+        )
+        for metavar in ('FILENAME', 'PATH')
+    ),
```
Clever.

witten commented on tests/unit/commands/test_completions.py:
```diff
@@ -0,0 +1,143 @@
+    has_unknown_required_param_options,
+)
+
+OptionType = namedtuple('OptionType', ['file', 'choice', 'unknown_required'])
```
I think this is a prime candidate for a `flexmock()` instance. It _should_ be a drop-in replacement. Although maybe the one thing you can't do easily with a flexmock instance is the `if True in ...` you have below.

isaec replied:
Why would I use a mock for this? My goal was to create a way to reuse the test cases between the different functions when some inputs would match for multiple functions--I used this tuple to replace having independent lists for each function that were ``(Action, Boolean)``
Is there a reason to mock it out?

witten replied:
Nope, I made this comment without realizing that these tuples aren't actually passed in anywhere. Given that they're just for test data structuring, I think they're fine as-is!

witten commented on tests/unit/commands/test_completions.py:
```diff
@@ -0,0 +1,143 @@
+def test_has_file_options_detects_file_options(action: Action, option_type: OptionType):
+    assert (
+        has_file_options(action) == option_type.file
+    ), f'Action: {action} should be file={option_type.file}'
```
IMO you might not need these messages. The default pytest assert introspection is pretty good these days and will produce a useful error message on its own. (Unless of course you've found otherwise...)

isaec replied:
sounds great, I have no experience with it, I just know assert doesn't produce nice messages in java and reached for an explicit message.

witten replied:
Feel free to play with it and decide for yourself (or not!)...

witten commented on tests/unit/commands/test_completions.py:
```diff
@@ -0,0 +1,143 @@
+
+
+@pytest.mark.parametrize('action, option_type', test_data)
+def test_produce_exact_options_completion(action: Action, option_type: OptionType):
```
Maybe remove the `produce_` here so it's clear what unit this is testing by the function name? You can always put more description after the test name if you need it. E.g. `test_exact_options_completion_produces_blah_blah_blah(...)`.

witten commented on tests/unit/commands/test_completions.py:
```diff
@@ -0,0 +1,143 @@
+        assert False, f'exact_options_completion raised ValueError: {value_error}'
+
+
+def test_dedent_strip_as_tuple():
```
I would add `_does_not_raise` to the end of this function name to make it abundantly clear what your test is doing.

witten reviewed (changes requested):
These tests look great! I really appreciate you taking the time to add them. I left a few minor comments for revision, but I think after that this is basically done. (Assuming you're willing for it to be. :smile:)

isaec commented:
I _hopefully_ didn't create new problems in fixing that, let me know if you have any more feedback.

witten reviewed (approved):
Awesome! Thanks again for sticking with this.

witten commented:
FYI: An unrelated ticket in borgmatic main (https://projects.torsion.org/borgmatic-collective/borgmatic/issues/697) adds support for borgmatic sub-actions, e.g. "borgmatic config bootstrap". This doesn't break the Fish shell completion, but it could work better with these sub-actions. For instance, "bootstrap" should probably not be a completion suggestion unless "config" is specified first, and "list" probably shouldn't be a suggestion after "borgmatic config". I've made these changes to the Bash completion but not to Fish. PRs welcome! :smiley:

isaec commented:
Thanks for the heads up. If I find myself using sub-actions, I'll take a look at improving things, but for now I think I'll leave this to another motivated user? :grin: It would be good to standardize things but I wonder if some of the specific nuances of when commands "make sense" and how you encode that would resist agnostic utility functions.

The current completions cover everything I use fully.
